resteasy-cdi/src/main/java/org/jboss/resteasy/cdi/ResteasyCdiExtension.java

```diff
@@ -81,9 +93,50 @@ public static boolean isCDIActive()
    public void observeBeforeBeanDiscovery(@Observes BeforeBeanDiscovery event, BeanManager beanManager)
    {
       this.beanManager = beanManager;
+      final AnnotatedType<ContextProducers> producersAnnotatedType = beanManager.createAnnotatedType(ContextProducers.class);
+      event.addAnnotatedType(producersAnnotatedType, ContextProducers.class.getCanonicalName());
       active = true;
    }
 
+   /**
+    * Process any client beans.
+    *
+    * @param pb the bean being processed
+    */
+   public void processClientBean(@Observes final ProcessBean<Client> pb) {
+      // this method will get notified if there is any bean of type `Client` created
+      final Set<Annotation> qualifiers = pb.getBean()
+              .getQualifiers(); // you want to detect beans with no explicit qualifiers
+      if (qualifiers.contains(Any.Literal.INSTANCE) && qualifiers.contains(Default.Literal.INSTANCE)) {
+         generateClientBean = false;
+      }
+   }
+
+   /**
+    * Process any producers for a client.
+    * @param producer the producer
+    * @param <T> the type of the bean class
+    */
+   public <T> void processClientProducers(@Observes final ProcessProducer<T, Client> producer) {
+      generateClientBean = false;
```

Comment:
It seems like here we might want to inspect the annotations to ensure there isn't some custom qualifier which produces it's own client.

Reply:
Note that `ProcessProducer` gets notified only for producer methods and field. Therefore, if user defined a class based bean with type `Client`, it won't trigger this observer. Same goes for a synthetic bean.